File: leet-code/509. Fibonacci Number/FibonacciNumber509_3.java
package com.example;

public class FibonacciNumber509_3 {

    public int fib(int n) {
        // 0 1
        // 1 1
        // 1 2
        // 2 3
        // 3 5
        // 5 8
        int x = 0, y = 1;
        for (int i = 0; i < n; i++) {
            int z = x + y;
            x = y;
            y = z;
        }
        return x;
    }
}

File: leet-code/509. Fibonacci Number/FibonacciNumber509_3.kt
package com.example

class FibonacciNumber509_3 {
    private val map = mutableMapOf(0 to 0, 1 to 1)

    fun fib(n: Int): Int {
        if (map.containsKey(n)) {
            return map[n]!!
        }
        map[n] = fib(n - 1) + fib(n - 2)
        return map[n]!!
    }
}
